RTL support for almost all registry components and examples

nahasco opened the pull request:
RTL support for almost all components in the registry except for table. 

Mainly replaced physical properties with logical properties and added `rtl:space-x-reverse` where `space-x-*` is used. Everything was done manually and reviewed carefully.

Also, fixed a bug where the default style sheet-side example was importing the new york style components instead of the default style components.

This fixes #530

nahasco commented:
#530

shadcn commented:
This is interesting. I wonder if we could have this as an option via the cli. Turn on `rtl` in components.json and we automatically apply those updates for you.

nahasco commented:
> This is interesting. I wonder if we could have this as an option via the cli. Turn on `rtl` in components.json and we automatically apply those updates for you.

That's a great idea!

The changes I have added result in support of both RTL and LTR directions, by simply adding `dir=rtl` to the html element `<html dir="rtl">` and finally wrapping the app with [Radix's direction provider](https://www.radix-ui.com/primitives/docs/utilities/direction-provider) `<DirectionProvider dir=rtl>` to switch to RTL. Furthermore, when changing the properties from directional to logical I realised some icons like `ChevronLeft` or `ChevronRight` also need to be adjusted for RTL.

Acheiving all of this via the cli would result in full RTL support, but I am not sure how can it be done.

shadcn commented:
@nahasco I can take a look at this. Having this in the cli would be awesome. Adding it to the next milestone.

nahasco commented:
> @nahasco I can take a look at this. Having this in the cli would be awesome. Adding it to the next milestone. I'll push to this PR and make sure you're credited for the work you've done here. Thank you.

Regarding the choice between physical and logical properties, I'd like to clarify whether you intend to use logical properties as the default for all languages. Logical properties offer the flexibility needed for seamless direction switching, and I believe they could simplify the codebase and improve adaptability for various languages.

For other specific elements like icons and HTML or radix direction values, there could be an option to manage these through the CLI to give a starting ground to the user. This approach allows us to make logical properties the primary choice for simplicity and flexibility while maintaining control over other elements that may require manual adjustments.

I'm curious to know your thoughts on this approach and whether logical properties are indeed the direction you're leaning towards. Your input on this would be highly valuable. Once again, thank you for your contribution to this!

tomer-yechiel commented:
please keep in mind the use case of supporting both LTR and RTL in the same codebase.

mustafamoe commented:
It would be great to have it rn 😅

dodeca-6-tope commented on apps/www/registry/default/ui/alert-dialog.tsx:
```diff
@@ -65,7 +65,7 @@ const AlertDialogFooter = ({
   <div
     className={cn(
-      "flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2",
+      "flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2 rtl:space-x-reverse",
```
Why not `flex flex-col-reverse sm:flex-row sm:justify-end sm:gap-2`?

nahasco replied:
I didnt want to change the original styling alot. Just adding rtl:space-x-reverse at the end  is the most minimal, i think.

MHBahrampour commented:
It's really shocking that a modern UI library doesn't fully support RTL or Internationalization (like calendar). Love shadcn but I with they would consider these stuff more seriously.

sahandsn commented:
any update on this?

salahkai commented:
Waiting for updates!

nahasco commented:
I will continue working on this in the next few days. I am hoping I add RTL support to as much components as I can before I open the PR again. Then we will have to wait for it to get merged. In the meantime, you may copy the RTL supported components from this PR.

nahasco commented:
Hey @shadcn , can you share some input on this? Any plans of merging?

shadcn commented:
@nahasco Yes. Let me add this to the roadmap. I have had this on my list for a while. This should be easier now I think with the new cli and registry.

nahasco commented:
Thank you @shadcn

> @nahasco question. is there a 1:1 mapping for those utility classes? I wonder if we could write a transformer to do these on install?


I am not sure what that means. But I didn't just adjust the classes, in some components I used radix ui's direction provider to pass the dir value to the component.

For example the carousel component requires a `dir` value to make it `rtl` or `ltr`.

I believe RTL support should be the default for the entire library and not just an option in the cli. The changes I made make the components support `rtl` and `ltr` dynamically based on the html `dir` value and radix's direction provider.

In a case where a user chooses not to support RTL in the cli, but later on decides to support RTL, they will have to refactor everything. So why not make RTL support the default?

iskaa02 commented:
My app support both RTL and LTR, For my usecase, I updated the code for RTL support by globally replacing `left-` with `start-` and `right-` with `end-`. This included properties like margins `ml` `mr` to `ms`  `me`, padding `pl` `pr` to `ps` `pe`, and text alignment `text-left` to `text-start`, ...etc.
also some bugs occurred from the translate-x class so for every `translate-x` i would append `rtl:-translate-x`.

For the sidebar, I just changed the side based on the lang

for dialogs, popovers, anything that would use a portal, i wrapped the children with a div with dir tag because it's not inherited from the root
```typescript
// src/components/ui/dialog.tsx
const DialogContent = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Content>
>(({ className, children, ...props }, ref) => (
  <DialogPortal>
    <div dir={languageTag() == "ar" ? "rtl" : "ltr"}>
      <DialogOverlay />
      <DialogPrimitive.Content
        ref={ref}
        className={cn(
          "fixed start-[50%] top-[50%] z-50 grid w-full max-w-lg rtl:translate-x-[50%] translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg",
          className,
        )}
        {...props}
      >
        {children}
        <DialogPrimitive.Close className="absolute end-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground">
          <Cross2Icon className="h-4 w-4" />
          <span className="sr-only">Close</span>
        </DialogPrimitive.Close>
      </DialogPrimitive.Content>
    </div>
  </DialogPortal>
));
```
Still facing small bugs here and there but nothing major

nitzanpap commented:
Subscribing to this!

jjeem commented:
Great work!
anything I can help with to push this ASAP?

nahasco commented:
> Great work!
> anything I can help with to push this ASAP?

Could you help with the remaining components?

omerman commented:
This is mazing following 🙏

nahasco commented:
> My app support both RTL and LTR, For my usecase, I updated the code for RTL support by globally replacing `left-` with `start-` and `right-` with `end-`. This included properties like margins `ml` `mr` to `ms`  `me`, padding `pl` `pr` to `ps` `pe`, and text alignment `text-left` to `text-start`, ...etc.
> also some bugs occurred from the translate-x class so for every `translate-x` i would append `rtl:-translate-x`.
> 
> For the sidebar, I just changed the side based on the lang
> 
> for dialogs, popovers, anything that would use a portal, i wrapped the children with a div with dir tag because it's not inherited from the root
> ```typescript
> // src/components/ui/dialog.tsx
> const DialogContent = React.forwardRef<
>   React.ElementRef<typeof DialogPrimitive.Content>,
>   React.ComponentPropsWithoutRef<typeof DialogPrimitive.Content>
> >(({ className, children, ...props }, ref) => (
>   <DialogPortal>
>     <div dir={languageTag() == "ar" ? "rtl" : "ltr"}>
>       <DialogOverlay />
>       <DialogPrimitive.Content
>         ref={ref}
>         className={cn(
>           "fixed start-[50%] top-[50%] z-50 grid w-full max-w-lg rtl:translate-x-[50%] translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg",
>           className,
>         )}
>         {...props}
>       >
>         {children}
>         <DialogPrimitive.Close className="absolute end-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground">
>           <Cross2Icon className="h-4 w-4" />
>           <span className="sr-only">Close</span>
>         </DialogPrimitive.Close>
>       </DialogPrimitive.Content>
>     </div>
>   </DialogPortal>
> ));
> ```
> Still facing small bugs here and there but nothing major

This is similar to what I did in this PR. Additionally, I also added a dir state to some components and added 'rtl:rotate-180' to left and right chevron icons to fully support RTL.

Aymen-Khanfir reviewed (approved):
Good job, i was thinking of use rlt[] and ltr[] but you solution required less code

nahasco commented:
@shadcn All components are completed and the PR is ready to merge.

shadcn commented:
@nahasco wow, great work. I'll review.

MohamedAlQadeery commented:
This is amazing cant wait for it be merged !

ioeldev commented:
Can we do something to speed up the merge ?

alialnaghmoush commented:
@shadcn We are waiting for the merged

200-0K commented:
Waiting for this to be merged

omerman commented:
Same here, would love to see it too

ioeldev commented:
If i have customized some of the shadcn components directly to the /ui folder (button.tsx, card.tsx etc.), how would i be able to keep my changes and integrate the new RTL support ?
Have someone already experienced a component update that has already custom style in it ?

omerman commented:
> If i have customized some of the shadcn components directly to the /ui folder (button.tsx, card.tsx etc.), how would i be able to keep my changes and integrate the new RTL support ? Have someone already experienced a component update that has already custom style in it ?

I guess you will have some conflicts.. hopefully not much 🤷

EmperorofWeb commented:
with  <html dir="rtl"> and DirectionProvider the sidebar widget still opens on left side

![image](https://github.com/user-attachments/assets/25f930b4-ca89-40eb-86b5-8db0c5a4caa4)

adotnusiyan commented:
@shadcn Please. We are waiting for this change to be merged.

b3of commented:
@shadcn Please. We are waiting for this change to be merged.

MRsabs commented:
any update? Can we do something to speed up the merge?

ilyassesalama commented:
@shadcn Still waiting for updates about this PR.

KINGJulienx commented:
meow (bump, looking forward to this merge 👀)

alan-ar1 commented:
Merge please 🙏

Jihad commented:
![image](https://github.com/user-attachments/assets/d0b0b886-2333-4d4d-8e77-d9b36eea9992)

ElyasSalar commented:
Waiting for merge 🍿

ioeldev commented:
Any blocker ? I need this also 🙏

LahmerIlyas commented:
What's the update on this?

ESSELAMI commented:
Had to change `left to start` and `right to end` and used i18n to translate the content and switching `dir`
https://github.com/user-attachments/assets/e468a517-43a0-46d8-aad3-f4f4e204e905
<img width="1280" alt="Screenshot 2025-02-28 at 12 12 47 PM" src="https://github.com/user-attachments/assets/af83cb3f-3985-4502-9577-dbc2f2c43789" />
<img width="1280" alt="Screenshot 2025-02-28 at 12 12 53 PM" src="https://github.com/user-attachments/assets/a98e56e1-e316-4ab9-986b-94c705aea3ef" />

seebham commented on apps/www/registry/default/ui/carousel.tsx:
```diff
@@ -19,6 +20,7 @@ type CarouselProps = {
   plugins?: CarouselPlugin
   orientation?: "horizontal" | "vertical"
   setApi?: (api: CarouselApi) => void
+  direction: "ltr" | "rtl"
```
Do we need this?

MarwanBz commented:
@nahasco @shadcn 
hey guys,
did merge comming or not?

AGI is coming and it's 2025 :D

nadavhalfon commented:
Any update?

jermanuts commented:
Any updates?

nahasco commented:
Waiting for an update from @shadcn

ItayElgazar commented:
@shadcn ? :)

speedydragon commented on apps/www/registry/default/ui/toaster.tsx:
```diff
@@ -9,12 +9,14 @@ import {
   ToastTitle,
   ToastViewport,
 } from "@/registry/default/ui/toast"
+import { useDirection } from "@radix-ui/react-direction"
```
can't resolve "@radix-ui/react-direction"

hmmftg replied:
it's a npm package: [link](https://www.npmjs.com/package/@radix-ui/react-direction)

davidulman commented:
+1

kamyabnazari commented:
Hello everyone, is there anything we can do to help, what else has to be done so RTL support would be added the Project?

Just wanting to ask, because I would love to have the possibility to use Shadcn-UI with RTL languages as well 👍🏼

mezzat11 commented:
<img width="220" height="225" alt="image" src="https://github.com/user-attachments/assets/510ce8e7-2e6e-4249-8dfb-c0a2f9846233" />
Hello @shadcn

iabdo9 commented:
temporary solution
https://github.com/shadcn-ui/ui/issues/2736#issuecomment-3009795255

Cyberistic commented:
Bumping and adding this here, as I believe using sides is the best approach:

https://github.com/m-farah/shadcn-rtl

https://github.com/m-farah/shadcn-rtl/blob/main/rtl-transform.ts

samerkh commented:
does this fix the concern mentioned [here](https://github.com/shadcn-ui/ui/pull/1635#issuecomment-1745294515)

mhmdsalahsebai commented:
Any update?

walidov commented:
@nahasco Hey, could you please resolve the conflicts, and let push @shadcn to merge this PR, it has been a while now and it is indeed very important to give shadcn-ui a boost by supporting RTL.

kamyabnazari commented:
Congratulations Everybody! 💯
